### ISSUE TITLE
[BUG] GM storage is not updating after page refresh in beta 2.15.8

### ISSUE THREAD

**Issue by Sv443** [violentmonkey/violentmonkey#1927]
Continuation from #1793

### Sequence of actions: 
1. Userscript is installed in a dev environment through a local webserver (I'm having this problem with my userscript [BetterYTM](https://github.com/Sv443/BetterYTM/blob/c909dde5699b6cbc047570b36c182a5dfa0a8a08/dist/BetterYTM.user.js))
2. music.youtube.com is opened
3. GM storage already contains the key `bytm-installed` from previous tests so no welcome menu pops up
4. All values in GM storage are cleared through a menu command using `GM.listValues()` and `GM.deleteValue()`
5. Listing all values in GM storage using a menu command yields no entries
6. The page is reloaded
7. The welcome menu still doesn't pop up. `GM.getValue("bytm-installed")` returns `{ ... }`. Running the "list values" menu command yields the entries that should've previously been deleted

The inverse can sometimes happen where GM storage gets "stuck" on empty values, so that no matter how many times `GM.setValue("bytm-installed", "{ ... }")` is run, the value of `GM.getValue("bytm-installed")` will always be `undefined`  
  
This problem does not occur on beta v2.15.5 and I haven't been able to reproduce it with a static script that was solely created for testing this issue so either this only happens with local development or with scripts that have at some point been installed already.  
I tested renaming the name and namespace in the header to no avail.  
  
I've recorded a video to help explain it: https://youtu.be/qcx-ryh4mJo  
Timestamps:  
- 0:00 script is installed, just got updated by the file being changed on the dev server  
- 0:04 page is reloaded  
- 0:08 initial value of `bytm-installed` (see sequence step 3)  
- 0:11 all values are deleted using [this menu command](https://github.com/Sv443/BetterYTM/blob/c909dde5699b6cbc047570b36c182a5dfa0a8a08/src/index.ts#L227) (see sequence step 4)  
- 0:16 page is reloaded, welcome menu should be shown  
- 0:20 GM.getValue() [here](https://github.com/Sv443/BetterYTM/blob/c909dde5699b6cbc047570b36c182a5dfa0a8a08/src/index.ts#L135) should return `undefined` since the storage was just cleared but returns old data (timestamp is exactly the same as before)  
- 0:26 storage is again deleted and listed to show that no values were found (sequence step 4 and 5)  
- 0:37 page is reloaded, welcome menu should be shown once again  
- 0:41 again, listing the values shows that `bytm-installed` is still present and contains the same exact timestamp  

Note: the values starting with `_uu` can be ignored for this example, though any updates to them are also discarded after page reload.

### Problem: 
See above.  
GM storage gets "stuck" on a previous set of values and doesn't apply any changes so they are persisted through page reloads.

### Expected result:  
GM storage should always persist.

#### Devtools console contents:
See video.

#### Environment:
- OS: MacOS 13.5.2
- Browser: Firefox Developer Edition 119.0b5
- Violentmonkey Version: Beta v2.15.8

**Comment by tophf** [comment]
It'd be helpful if you test the [latest CI build](https://github.com/violentmonkey/violentmonkey/actions/workflows/ci.yml) so that I don't make another bugged beta.

**Comment by Sv443** [comment]
Tested with v2.15.8-12-gdbc0a79 and unfortunately I still have the same issue.  
The values are only persisted after a complete browser restart, but while the session is active it keeps rolling back to the original data, no matter how many times the page is reloaded or reopened.

**Comment by Sv443** [comment]
I created a new branch [`vm_bug`](https://github.com/Sv443/BetterYTM/blob/vm_bug/dist/BetterYTM.user.js) which has all features disabled except the welcome menu I've mentioned, also all the URLs are corrected. If you checked it out you can run `npm run dev` to host it locally like I did.

**Comment by tophf** [comment]
Nothing's changed for me, no welcome menu, the same messages (except the build number) and errors are shown in the console.

**Comment by tophf** [comment]
I've set a breakpoint inside our injected.js at `ScriptData` and I see that your script receives an empty storage after I delete the values via the menu command and reload the tab.

**Comment by Sv443** [comment]
> I've set a breakpoint inside our injected.js at `ScriptData` and I see that your script receives an empty storage after I delete the values via the menu command and reload the tab.

Yes initially it should be empty. Only after it has been set and the entire browser is restarted does it give my script the correct storage values.

**Comment by tophf** [comment]
My point is that it becomes empty after I delete its current contents via the menu command. This is the correct behavior, which I also see in any other script, including yours.

**Comment by Sv443** [comment]
I just pushed another commit which doesn't have any `@resource` directives and their associated code so there shouldn't be any errors regarding loading translations. It should literally just be the welcome menu thing and nothing else.

**Comment by Sv443** [comment]
> My point is that it becomes empty after I delete its current contents via the menu command. This is the correct behavior, which I also see in any other script, including yours.

So if at the point of loading the tab the storage is filled with values, I use the command to delete all values, it will show them as deleted for as long as the tab is not reloaded. But as soon as it's reloaded for me it gets filled up again with the initial values, even though no code is called to fill it with those values.

**Comment by tophf** [comment]
Here it doesn't get refilled. The new version shows the menu on each tab reload, I'll investigate what happens...

**Comment by Sv443** [comment]
Yeah so if you close the welcome menu, it should create the `bytm-installed` value in GM storage and never show that menu again on reload. But on my end it does, since the storage is rolled back to being empty. Only after a complete browser restart while the tab is active does it keep the values.

**Comment by tophf** [comment]
Yeah I've reproduced it.

**Comment by tophf** [comment]
Now I can't replicate it anymore... I have a general idea of what's wrong though.

**Comment by Sv443** [comment]
Just tested it with [this build](https://github.com/violentmonkey/violentmonkey/actions/runs/6484648028) and unfortunately it's still not fixed on my end, even after a fresh install of the extension and userscript and restarting the browser in between.

**Comment by Sv443** [comment]
I literally just run `npm run dev` and open the link that's shown in the console, select "track external edits" and install the script.  
I did just now test the latest CI build on Firefox Developer, Firefox Nightly and Chrome Stable and it only happens on Firefox. On Chrome I got the same errors you did initially, then when I switched to the `vm_bug` branch the GM storage works just as it should. So this has to be some kind of issue with Firefox specifically.  
I also tried running VM as the only extension with incognito mode privileges to see whether it was another extension messing it up but that didn't make a difference.
